src/awkward/operations/ak_concatenate.py

```diff
@@ -143,23 +155,25 @@ def action(inputs, depth, **kwargs):
                         nextinputs.append(x)
                 inputs = nextinputs
 
-            if depth == posaxis:
-                backend = backend_of(*inputs, default=cpu)
-
-                length = None
-                for x in inputs:
-                    if isinstance(x, ak.contents.Content):
-                        if not is_integer(length):
-                            length = x.length
-                        elif length != x.length and is_integer(x.length):
-                            raise ValueError(
-                                "all arrays must have the same length for "
-                                "axis={}".format(axis)
-                            )
+            # Ensure the lengths agree, taking known lengths over unknown lengths
+            length = None
+            for x in inputs:
+                if isinstance(x, ak.contents.Content):
+                    if length is None:
+                        length = x.length
+                    elif x.length is unknown_length:
+                        continue
+                    elif length is unknown_length:
+                        length = x.length
+                    elif length != x.length:
+                        raise ValueError(
+                            "all arrays must have the same length for "
+                            "axis={}".format(axis)
+                        )
+            assert length is not None
```

Comment:
Yes. If, at this stage, you already know there will be an error because at least two lengths are known and conflict, then it's time to raise an error.